include/natalie/module_object.hpp

```diff
@@ -164,6 +173,8 @@ class ModuleObject : public Object {
     Vector<ModuleObject *> m_included_modules {};
     MethodVisibility m_method_visibility { MethodVisibility::Public };
     bool m_module_function { false };
+
+    TM::Hashmap<SymbolObject *> m_non_overwritable_methods {};
```

Comment:
I don't like that we need another Hashmap on every module, but it seems to have negligible effect on performance from my testing locally.

Still, if I could think of a way to do this without a bunch of hashmaps, I would. Perhaps a single Hashmap on GlobalEnv that uses a `std::pair<ModuleObject*, SymbolObject*>` for its key... :thinking:

Reply:
> Still, if I could think of a way to do this without a bunch of hashmaps, I would. Perhaps a single Hashmap on GlobalEnv that uses a `std::pair<ModuleObject*, SymbolObject*>` for its key... 🤔

I've tried implementing this but it seems like our HashMap implementation does not support hashing std::pair objects. I'm getting the following error:
```
natalie/include/tm/hashmap.hpp:152:31: error: cannot convert ‘std::pair<Natalie::ModuleObject*, Natalie::SymbolObject*>’ to ‘const void*’ in argument passing
  152 |         auto hash = m_hash_fn(key);
      |                               ^~~
      |                               |
      |                               std::pair<Natalie::ModuleObject*, Natalie::SymbolObject*>
```

As a C++ beginner I'm not quite sure why one HashMap for each ModuleObject would result in a big performance decrease. Is it because of the initialization? Maybe an alternative would be to set the initial capacity of those HashMaps to 0?

Reply:
I just checked the Hashmap code, and it seems we do not allocate the chunk of memory for storing items until the first item is added. In that case, I would feel ok with this as-is.